enhancements/monitoring/integration-with-topology-modes.md

```diff
@@ -0,0 +1,90 @@
+---
+title: monitoring-topology-mode
+authors:
+  - "@lilic"
+reviewers:
+  - "monitoring-team"
+  - "???"
+approvers:
+  - "monitoring-team"
+creation-date: 2021-03-01
+last-updated: 2021-03-01
+status: implementable
+see-also:
+  - "/enhancements/cluster-high-availability-mode-api.md"
+---
+
+# Monitoring Stack on Single Replica Topology Mode
+
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The monitoring stack needs to react based on the `infrastructureTopology` field newly added to `infrastructures.config.openshift.io` CRD. Plan is to detect whenever
+cluster is in single replica topology mode and reduce the monitoring stack footprint by adjusting platform monitoring
+instances to be non Highly Available.
+
+## Motivation
+
+In the 4.8 time frame, OpenShift will introduce a new capabilities field in the Infrastructure API that will inform
+operators about the deployment topology. The immediate goal is for operators to adjust the deployment of their operands
+depending on whether the infrastructure is HA (at least 3 nodes) or not (single node).
+
+### Goals
+
+- Reduce monitoring footprint by lowering number of replicas
+
+### Non-Goals
+
+- User workload stack or its Prometheus instance is not adjusted
+- Reduce memory footprint
+- Address CRC requirements (though it might help)
+- Tests from monitoring team
+
+## Proposal
+
+To adjust to the single replica mode, in cluster-monitoring-operator we detect the topology mode based on the "cluster"
+named instance of `Infrastructure`, if that mode is `SingleReplicaTopologyMode`, we only spin up one instance of
+platform Prometheus for that cluster.
+
+### Risks and Mitigations
+
+There are risks that this is not enough to reduce the actual footprint of the monitoring stack. The other thing is that
+as a result of this monitoring is no longer highly available but there is no workaround for that as its a single node
+cluster.
+
+## Design Details
+
+### Test Plan
+
+N/A Its a non goal for monitoring team, apart from manual testing.
+
+### Graduation Criteria
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+Upgrades are a non goal right now.
+
+### Version Skew Strategy
+
+N/A
```

Comment:
The new monitoring code needs to default to its current HA choices when it does not find `infrastructureTopology` on the status, so we don't break existing clusters when they update into the new monitoring operator code.  I think someone may have confirmed that the client injects [the kubebuilder default][1] automatically, but I don't remember clearly enough to feel comfortable with a N/A here.  Can we at least say that we need new monitoring operator code to be compatible with `Infrastructure` where the new topology fields are unset, preserving the implicit HA semantics, even if we don't spell out how that compatibility is going to work at the code level?

[1]: https://github.com/openshift/api/blob/175513fcccbe4eb5223cb45a7246bf697b835c3e/config/v1/types_infrastructure.go#L94

Reply:
FWIW we've already validated that the field defaults to `HighlyAvailable`. But agreed that we can even ensure that the cluster monitoring operator will fallback to the default HA deployment if the `infrastructureTopology` field is missing.